File: examples/java/Reverse.java
/**
 * Reverse-geocode a single point with RevAddr (Java 11+ HttpClient).
 *
 * Compile and run:
 *   export REVADDR_API_KEY=sk_live_...
 *   javac Reverse.java
 *   java Reverse
 *   java Reverse 37.7749 -122.4194
 *
 * How it works: GET /v1/reverse with x-api-key, parse JSON for result.formatted_address.
 * This sample prints the full body; for production prefer a JSON library (Jackson/Gson).
 */

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;

public class Reverse {
    private static final String DEFAULT_BASE = "https://api.revaddr.com";

    public static void main(String[] args) throws Exception {
        String apiKey = System.getenv("REVADDR_API_KEY");
        if (apiKey == null || apiKey.isBlank()) {
            System.err.println("Set REVADDR_API_KEY to your RevAddr API key.");
            System.exit(1);
        }

        String base = System.getenv("REVADDR_BASE_URL");
        if (base == null || base.isBlank()) {
            base = DEFAULT_BASE;
        }

        // Defaults match the website hero example.
        double lat = 38.8977;
        double lon = -77.0365;
        if (args.length >= 2) {
            lat = Double.parseDouble(args[0]);
            lon = Double.parseDouble(args[1]);
        }

        // Build query string with UTF-8 encoding (Invariant decimal form is fine for doubles).
        String query =
            "lat=" + URLEncoder.encode(Double.toString(lat), StandardCharsets.UTF_8)
                + "&lon=" + URLEncoder.encode(Double.toString(lon), StandardCharsets.UTF_8);
        URI uri = URI.create(base.replaceAll("/$", "") + "/v1/reverse?" + query);

        HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();

        // Auth and Accept headers on every reverse call.
        HttpRequest request = HttpRequest.newBuilder(uri)
            .timeout(Duration.ofSeconds(30))
            .header("x-api-key", apiKey.trim())
            .header("Accept", "application/json")
            .GET()
            .build();

        HttpResponse<String> response =
            client.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            System.err.println("HTTP " + response.statusCode() + ": " + response.body());
            System.exit(1);
        }

        // Full JSON: {"result":{"formatted_address":"...", ...}}
        System.out.println(response.body());
    }
}

File: examples/kotlin/Reverse.kt
/**
 * Reverse-geocode a single lat/lon with RevAddr (Kotlin, JVM).
 *
 * Requirements: JDK 11+, Kotlin compiler. No third-party deps (java.net.http).
 *
 * Run:
 *   export REVADDR_API_KEY=sk_live_...
 *   kotlinc Reverse.kt -include-runtime -d reverse.jar
 *   java -jar reverse.jar
 *   java -jar reverse.jar 37.7749 -122.4194
 *
 * Android/backend: copy reverseGeocode() into a repository class and load the
 * API key from secrets (never commit keys; avoid logging full keys).
 */

import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.time.Duration
import kotlin.system.exitProcess

const val DEFAULT_BASE = "https://api.revaddr.com"

/**
 * GET /v1/reverse and return the raw JSON body.
 * Production apps should map fields into a data class (kotlinx.serialization / Moshi).
 */
fun reverseGeocode(
    lat: Double,
    lon: Double,
    apiKey: String,
    baseUrl: String = DEFAULT_BASE,
): String {
    require(apiKey.isNotBlank()) { "apiKey is required" }

    val base = baseUrl.trimEnd('/')
    val query =
        "lat=" + URLEncoder.encode(lat.toString(), StandardCharsets.UTF_8) +
            "&lon=" + URLEncoder.encode(lon.toString(), StandardCharsets.UTF_8)
    val uri = URI.create("$base/v1/reverse?$query")

    val client =
        HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build()

    // Auth is always the x-api-key header (never a query param).
    val request =
        HttpRequest.newBuilder(uri)
            .timeout(Duration.ofSeconds(30))
            .header("x-api-key", apiKey)
            .header("Accept", "application/json")
            .GET()
            .build()

    val response = client.send(request, HttpResponse.BodyHandlers.ofString())
    if (response.statusCode() !in 200..299) {
        throw RuntimeException("HTTP ${response.statusCode()}: ${response.body()}")
    }
    return response.body()
}

/** Minimal pull of result.formatted_address without a JSON library (demo only). */
fun formattedAddressFromBody(body: String): String? {
    val key = "\"formatted_address\""
    val i = body.indexOf(key)
    if (i < 0) return null
    val colon = body.indexOf(':', i + key.length)
    val firstQuote = body.indexOf('"', colon + 1)
    if (firstQuote < 0) return null
    val secondQuote = body.indexOf('"', firstQuote + 1)
    if (secondQuote < 0) return null
    return body.substring(firstQuote + 1, secondQuote)
}

fun main(args: Array<String>) {
    val apiKey = System.getenv("REVADDR_API_KEY")?.trim().orEmpty()
    if (apiKey.isEmpty()) {
        System.err.println("Set REVADDR_API_KEY to your RevAddr API key.")
        exitProcess(1)
    }
    val base = System.getenv("REVADDR_BASE_URL")?.trim()?.ifEmpty { null } ?: DEFAULT_BASE

    // Defaults match the website hero example (White House).
    val lat = args.getOrNull(0)?.toDoubleOrNull() ?: 38.8977
    val lon = args.getOrNull(1)?.toDoubleOrNull() ?: -77.0365

    try {
        val body = reverseGeocode(lat, lon, apiKey, base)
        println(formattedAddressFromBody(body) ?: "(no formatted_address)")
        println(body)
    } catch (e: Exception) {
        System.err.println(e.message)
        exitProcess(1)
    }
}
